traefik/Changelog.md

```diff
@@ -1,5 +1,57 @@
 # Change Log
 
+## 18.0.0 
+
+**Release date:** 2022-10-21
+
+![AppVersion: 2.9.1](https://img.shields.io/static/v1?label=AppVersion&message=2.9.1&color=success&logo=)
+![Helm: v3](https://img.shields.io/static/v1?label=Helm&message=v3&color=informational&logo=helm)
+
+
+* Update Changelog and Chart version 
+* Add additional test for exposing http3 without enabling tls 
+* Split service tests based on single vs split service flag 
+* Merge TCP and UDP ports into single service by default 
+* Add service-ports helper to _service.tpl 
+
+### Default value changes
+
+```diff
+diff --git a/traefik/values.yaml b/traefik/values.yaml
+index 807bd09..fc5aff3 100644
+--- a/traefik/values.yaml
++++ b/traefik/values.yaml
+@@ -421,10 +421,16 @@ ports:
+     # The port protocol (TCP/UDP)
+     protocol: TCP
+     # nodePort: 32443
+-    # Enable HTTP/3.
+-    # Requires enabling experimental http3 feature and tls.
+-    # Note that you cannot have a UDP entrypoint with the same port.
+-    # http3: true
++    http3:
++      port: 8443
++      # hostPort: 8443
++      # Enable HTTP/3.
++      # Requires enabling experimental http3 feature and tls.
++      expose: false
++      # Note that you cannot have another UDP entrypoint with the same (exposed) port.
++      exposedPort: 8443
++      # The port protocol (TCP/UDP)
++      protocol: UDP
+     # Set TLS at the entrypoint
+     # https://doc.traefik.io/traefik/routing/entrypoints/#tls
+     tls:
+@@ -500,6 +506,7 @@ tlsStore: {}
+ # from.
+ service:
+   enabled: true
++  single: true
```

Comment:
Yeah, I had some doubts about it myself as well. Basically my argumentation is: this default seems to be both the most-used and most-reasonable value. 

The old value creates 2 separate services, which will result in 2 different LBs. Although some people might actually want to do that, I think most people starting out with this chart will be utterly confused (like me) that enabling HTTP/3 causes an additional LB to spin up. Moreover, they'll probably revert in this case, as routing HTTP/3 through a different LB is even more advanced and more costly as well.

With the new value, people trying to enable HTTP/3 on a non-supporting cluster will simply run into a Kubernetes warning/error telling them there is no support for their service, which feels more 'logical' to me.

You could consider adding some sort of service-detection for multi-protocol support, and warning users in advance.

Reply:
Also note that the multi-protocol feature has been available since Kubernetes 1.20 (see https://github.com/kubernetes/kubernetes/pull/94028) behind a feature-flag; it graduated to beta in 1.24. So there may be more providers who support this.

Reply:
You're right, it's possible to enable it since v1.20 and it's not enabled by default before v1.24.
It's still an advanced feature that is available but needs to be enabled on most cloud & kubernetes providers.
Http3 is also an experimental feature. 

Following your arguments, I was thinking about doing both, I mean
1. Disabling single service by default
2. Enabling single service when user enable `http3`

But it means that it won't be possible to use `http3` with two services.
And that's how one can do it currently with this Chart. 
And it works on most cloud & kubernetes providers.

=> So, for this year, I guess we should disable it by default and improve documentation on http3 about this.

Reply:
I was going to open a ticket about something related but noticed this WIP. I like to add that in some providers you need a TCP port open in the NLB for the health check. I came across this because there is no way of deploying the UDP NLB with this chart for 'digitalocean' where you will get this error message "Error syncing load balancer: failed to ensure load balancer: failed to build load-balancer request: specified health check port 80 does not exist on service traefik/traefik-udp". So, it seems odd that to get my single service, and single NLB, I need to enable `http3`. Seems unrelated otherwise.

Reply:
Just for completion. If you attempt to use the same port you opened with UDP protocol you get the following error "Error syncing load balancer: failed to ensure load balancer: failed to build load-balancer request: health check port: 4648, cannot be of protocol type UDP". So, this ticket will solve more than optimising the services into one. For some providers will make it possible to use UDP inbound.

Reply:
> You're right, it's possible to enable it since v1.20 and it's not enabled by default before v1.24. It's still an advanced feature that is available but needs to be enabled on most cloud & kubernetes providers. Http3 is also an experimental feature.
> 
> Following your arguments, I was thinking about doing both, I mean
> 
> 1. Disabling single service by default
> 2. Enabling single service when user enable `http3`
> 
> But it means that it won't be possible to use `http3` with two services. And that's how one can do it currently with this Chart. And it works on most cloud & kubernetes providers.
> 
> => So, for this year, I guess we should disable it by default and improve documentation on http3 about this.

I do agree support for mixed protocols might be limited (you could actually do a quick check to see how well-supported it is among major providers; I'm using Digital Ocean and they only added UDP support recently), so that might be a reason to keep the old default.

Then again, I think the current default of automatically creating two LB's is even more advanced. My assumption is few people are using this in practice, as it will mean 2 IP's and hence a lot of hassle to use. Actually I'm still curious if there are any user stories of people using the current chart for http3, hence my initial call for feedback.

Moreover the result of suddenly having two load balancers is confusing and potentially dangerous (I ended up accidentally deleting my LB while rollbacking my chart the first time I tried).

So basically I think the old default only already serves quite advanced (edge) cases. 

In the end it's your call what's best for now, happy to change back to the old value if requested.

Side note: I will probably not have any time to adjust things this week, so I'll look further into any requested changes next week.

Reply:
It must be just dumb easy. One Boolean value more or less - is fine. I can recall traefik in its early days already supporting let's encrypt and http2 and that was why we chose it. I tried to configure traefik with MetalLB and two services last week with current chart and I literally was reading the chart sources to find out what to configure and in the end failed. I'm not an advanced networker nor sysadmin, I must admit; however just to enable http3 it is too much. E.g. compare traefiks docker-compose http3 example (where one needs to just open the port) and what one needs to do with the helm chart

Reply:
@jnoordsij it looks like you are suggesting that there is a way to deploy this chart in Digital Ocean. I don't think that this is possible when you are exposing TCP and UDP ports. This chart selects all and any UDP port for a traefik-udp port and doesn't let us add a TCP port in that LB for healthcheck. I ended up cloning this chart locally and modifying the service template which is sad but while we are conversing here which seems will take a few weeks to sort out.

Reply:
> @jnoordsij it looks like you are suggesting that there is a way to deploy this chart in Digital Ocean. I don't think that this is possible when you are exposing TCP and UDP ports. This chart selects all and any UDP port for a traefik-udp port and doesn't let us add a TCP port in that LB for healthcheck. I ended up cloning this chart locally and modifying the service template which is sad but while we are conversing here which seems will take a few weeks to sort out.

You're totally right, my apologies if my comments were unclear. The latest release of this chart does not support HTTP/3 with Digital Ocean setup 'out-of-the-box', that's why I started this MR in the first place. Using my local version of this chart with the proposed changes, I did manage to get it working now.